### ISSUE TITLE
TcpListener::from_std is super easy to misuse

### ISSUE THREAD

**Issue by Ten0** [tokio-rs/tokio#5595]
**Is your feature request related to a problem? Please describe.**
I lost about 6 hours tracking down a deadlock bug that ended up leading me to this in `TcpListener::from_std`:
https://github.com/tokio-rs/tokio/blob/88445e762c569b538ac3f8d3dbb76887db421207/tokio/src/net/tcp/listener.rs#L206-L209

I would argue that this API is very prone to misuse, especially since going through it is the only way to attempt binding an address outside of tokio then turn it into a tokio-usable one (even in the context of hyper/axum).

**Describe the solution you'd like**
I'd like the public API for this feature of converting from STD to tokio to make the necessary ` std_listener.set_nonblocking(true)` call.

Alternately, I think if we have a function that does not make that call, it should be named in a way that makes it explicit that the user has to do that specific thing that's not enforced by typing before calling it, e.g. `TcpListener::from_std_manually_configured_as_non_blocking`

**Additional context**
It seems that [this used to be the way this worked](https://github.com/tokio-rs/tokio/issues/3001), but was changed propagating a change from downstream.
I would tend to think that it may have been best to not propagate that change directly, and instead do one of the two things above.

**Comment by Darksonn** [comment]
Yes, I agree, this has been a problem for a while. We can't change the existing method, but maybe we could deprecate it and replace it with `from_tcp` and `from_tcp_unchecked` where the first method sets is as non-blocking and checks that it's actually a tcp stream. This would be similar to `from_file` on pipes.

**Comment by Ten0** [comment]
> maybe we could deprecate it and replace it with from_tcp and from_tcp_unchecked where the first method sets is as non-blocking and checks that it's actually a tcp stream. This would be similar to from_file on pipes.

That sounds like a great idea. Should I open a PR?

**Comment by Ten0** [comment]
I'm not sure what you mean by "checks that it's actually a tcp stream" though: the entry type is [`TcpListener`](https://doc.rust-lang.org/stable/std/net/struct.TcpListener.html) so AFAICT that in particular is already guaranteed by typing, no?

**Comment by Ten0** [comment]
That seems to be checking that the file is a [fifo](https://man7.org/linux/man-pages/man7/fifo.7.html), which seems to not be guaranteed by typing there.

(Just to make sure it wasn't missed: Should I open a PR?)

**Comment by Darksonn** [comment]
Yes, go ahead and open a PR.

**Comment by tsar-boomba** [comment]
Just fixed a bug from 2 years ago caused by this. Would love to get this merged in ASAP. Will do anything to help out.

**Comment by rupert648** [comment]
I was just caught out by this as well 😅 
Is there anything we can do to push this forward?

**Comment by Noah-Kennedy** [comment]
@carllerche @Darksonn this continues to cause folks real production incidents, and is by far the most dangerous footgun lurking in Tokio today.

I think it's time we revisited whether fixing this is really a breaking change, and prepare a plan to deprecate all existing APIs that have this behavior if we cannot fix this.

AFAICT this seems to have been [the behavior in Tokio 0.2.x](https://docs.rs/tokio/0.2.25/tokio/net/struct.TcpListener.html#method.from_std). I do not know why we changed it, but I'd argue that this current behavior is clearly a bug.

I'm all for creating an unsafe or "discouraged" interface that gives users a blocking socket and has plenty of warnings, but I think the current behavior needs to go.

I think this is no more a breaking change than changing how we set the number of worker threads was. Both are behavioral changes which bring Tokio closer into alignment with how users think our system works, and both fix repeated causes of serious incidents which our users need to constantly guard against, and both behavioral changes are for things where there is no legitimate reason to depend on the current behavior. Everyone who hits the current behavior views this as a bug and is upset because they wasted their time, or worse, had a production incident.

If we cannot change this then we should deprecate the existing APIs (not sure what to do with TryFrom though) and have new ones which do the right thing that we can tell folks to use.

**Comment by pepyakin** [comment]
FWIW, I just got caught by the same but unlike other posters I managed to do it with a UDS.

**Comment by ndevenish** [comment]
I just spent several hours tracking down an issue where TcpListener would apparently block. It wasn't exactly this: It was the exact same issue on `UdpListener`, that the UdpListener build with `from_std` (so it could set_reuse_port) was blocking the thread.

Is it worth updating `UdpListener::from_std` for the same issue?
